src/common/cockpitconf.c

```diff
@@ -115,188 +175,202 @@ cockpit_conf_init (void)
+          env = strdup (env);
+          for (i = 0, token = strtok (env, ":");
+               token;
+               i++, token = strtok (NULL, ":"))
+            {
+              system_config_dirs = reallocarr (system_config_dirs, i + 2, sizeof (char*));
+              system_config_dirs[i] = strdup (token);
+            }
+          system_config_dirs[i] = NULL;
+          free (env);
         }
     }
 
-  g_free (cmp);
-  return ret;
+  return system_config_dirs ?: cockpit_config_dirs;
 }
 
-static const gchar **
-build_strv (const gchar *section,
-            const gchar *field,
-            gchar *delimiter)
+static Entry*
+cockpit_conf_lookup (const char *section,
+                     const char *field)
 {
-  const gchar *string = NULL;
-  const gchar **value = NULL;
-  gchar *stripped = NULL;
-  gchar **strv = NULL;
+  Entry *e;
+  char *section_lower, *field_lower;
 
-  string = cockpit_conf_string (section, field);
+  if (section == NULL || field == NULL)
+    return NULL;
 
-  if (string != NULL)
-    {
-      stripped = g_strstrip (g_strdup (string));
-      strv = g_strsplit (stripped, delimiter, -1);
-      if (strv)
-        {
-          guint n = g_strv_length (strv);
-          guint i;
-          value = g_new (const gchar *, n + 1);
+  ensure_cockpit_conf ();
 
-          for (i = 0; i < n; i++)
-            value[i] = g_strdup (strv[i]);
+  section_lower = strlower (strdup (section));
+  field_lower = strlower (strdup (field));
 
-          value[i] = NULL;
-          g_strfreev (strv);
-        }
+  for (e = cockpit_conf; e; e = e->next)
+    {
+      if (strcmp (e->section, section_lower) == 0 && strcmp (e->key, field_lower) == 0)
+        break;
     }
-  g_free (stripped);
-  return value;
+
+  free (section_lower);
+  free (field_lower);
+  return e;
 }
 
-const gchar **
-cockpit_conf_strv (const gchar *section,
-                   const gchar *field,
-                   gchar delimiter)
+const char *
+cockpit_conf_string (const char *section,
+                     const char *field)
 {
-  const gchar **value = NULL;
-  gchar *key = NULL;
-  gchar delm[2] = {delimiter, '\0'};
+  const Entry *e = cockpit_conf_lookup (section, field);
+  return e ? e->value : NULL;
+}
 
-  g_return_val_if_fail (section != NULL, NULL);
-  g_return_val_if_fail (field != NULL, NULL);
+bool
+cockpit_conf_bool (const char *section,
+                   const char *field,
+                   bool defawlt)
+{
+  const char *value = cockpit_conf_string (section, field);
+  if (value)
+    return strcasecmp (value, "yes") == 0 || strcasecmp (value, "true") == 0 || strcmp (value, "1") == 0;
+  return defawlt;
+}
 
-  ensure_cockpit_conf ();
-  key = g_strjoin (delm, section, field, NULL);
+static const char **
+get_strv (Entry *entry, char delimiter)
+{
+  char *stripped, *token;
+  char delimiters[2] = "x";
+  unsigned i;
+
+  if (!entry->value)
+    return NULL;
 
-  if (!g_hash_table_contains (cached_strvs, key))
+  if (entry->strv_cache)
```

Comment:
Do we care about look ups from multiple threads?  If so, we need to make this a critical section.

Reply:
We don't use threads in Cockpit. Our code in master doesn't use any locking either. If we ever need that, then IMHO we can add this in a separate PR, but that's going to be a huge amount of work (across all of cockpit, I mean)